src/Middleware/RequestAutoPopulationMiddleware.php

```diff
@@ -0,0 +1,103 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+namespace Google\ApiCore\Middleware;
+
+use Google\ApiCore\Call;
+use GuzzleHttp\Promise\PromiseInterface;
+use Ramsey\Uuid\Uuid;
+
+/**
+ * Middleware that adds autopopulation functionality. This middlware is
+ * added iff auto population settings are present in the resource
+ * descriptor config for the rpc method in context.
+ *
+ * @internal
+ */
+class RequestAutoPopulationMiddleware implements MiddlewareInterface
+{
+    /** @var callable */
+    private $nextHandler;
+
+    /** @var array<string, string> */
+    private $autoPopulationSettings;
+
+    public function __construct(
+        callable $nextHandler,
+        array $autoPopulationSettings
+    ) {
+        $this->nextHandler = $nextHandler;
+        $this->autoPopulationSettings = $autoPopulationSettings;
+    }
+
+    /**
+     * @param Call $call
+     * @param array $options
+     *
+     * @return PromiseInterface
+     */
+    public function __invoke(Call $call, array $options)
+    {
+        $next = $this->nextHandler;
+
+        if (empty($this->autoPopulationSettings)) {
+            return $next($call, $options);
+        }
+
+        $request = $call->getMessage();
+        foreach ($this->autoPopulationSettings as $fieldName => $valueType) {
+            $getFieldName = 'get' . ucwords($fieldName);
+            // We use a getter instead of a hazzer here because there's no need to
+            // differentiate between isset and an empty default value. Even if a
+            // field is explicitly set to an empty string, we want to autopopulate it.
+            if (empty($request->$getFieldName())) {
+                $setFieldName = 'set' . ucwords($fieldName);
+                switch ($valueType) {
+                    case \Google\Api\FieldInfo\Format::UUID4:
+                        $request->$setFieldName(Uuid::uuid4()->toString());
+                        break;
+                    default:
+                        throw new \UnexpectedValueException(sprintf(
+                            "Value type %s::%s not supported for auto population of the field %s",
+                            \Google\Api\FieldInfo\Format::class,
+                            \Google\Api\FieldInfo\Format::name($valueType),
```

Comment:
nit, import this so we don't have to use absolute paths

Reply:
Done 🚀